cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,317 @@
+ *
+ * NOTE: There is no need to invalidate the cache at any point since we will always need to call appfabric to confirm
+ * that the cached version is the newest version available.
+ */
+public class ArtifactLocalizer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactLocalizer.class);
+  // TODO: Move this directory name into a cConf property
+
+  private static final String LAST_MODIFIED_HEADER = HttpHeaderNames.LAST_MODIFIED.toLowerCase();
+
+  private final RemoteClient remoteClient;
+  private final RetryStrategy retryStrategy;
+  private final String dataDir;
+
+  @Inject
+  ArtifactLocalizer(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
+                                         HttpRequestConfig.DEFAULT,
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+    this.dataDir = cConf.get(CFG_LOCAL_DATA_DIR);
+  }
+
+  /**
+   * Gets the location on the local filesystem for the given artifact. This method handles fetching the artifact as well
+   * as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @return The Local Location for this artifact
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Fetching artifact info for {}", artifactId);
+    return Retries.callWithRetries(() -> fetchArtifact(artifactId), retryStrategy);
+  }
+
+  /**
+   * Gets the location on the local filesystem for the directory that contains the unpacked artifact. This method
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    File jarLocation = getArtifact(artifactId);
+    File unpackDir = getUnpackLocalPath(artifactId, Long.parseLong(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesn't exist, unpacking into {}", unpackDir);
+      BundleJarUtil.unJar(jarLocation, unpackDir);
```

Comment:
Is this method also thread safe?

Reply:
ah good point, added the same temp-file logic as the fetchArtifact method to make it thread safe